### PR TITLE
A bit of ReferenceManager refactoring

### PR DESCRIPTION
**Pull request by tmat**
Avoid potential non-determinism (enumerating hashtables), use pooled ArrayBuilders, expression-bodied properties, unify C# and VB.

@AlekseyTs @VSadov

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 1]
```diff
@@ -978,33 +978,50 @@ public override CompilationReference ToMetadataReference(ImmutableArray<string>
...
+            modules.AddRange(Assembly.Modules);
+
+            var referenceManager = GetBoundReferenceManager();
 
-            foreach (var pair in GetBoundReferenceManager().ReferencedAssembliesMap)
+            for (int i = 0; i < referenceManager.ReferencedAssemblies.Length; i++)
             {
-                MetadataReference reference = pair.Key;
-                ReferenceManager.ReferencedAssembly referencedAssembly = pair.Value;
-                if (reference.Properties.Kind == MetadataImageKind.Assembly) // Already handled modules above.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: Is this condition not needed now?

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 2]
```diff
@@ -978,33 +978,50 @@ public override CompilationReference ToMetadataReference(ImmutableArray<string>
...
-            return modules;
+        /// <summary>
+        /// Return a list of assembly symbols than can be accessed without using an alias.
+        /// For example:
+        ///   1) /r:A.dll /r:B.dll -> A, B
+        ///   2) /r:Foo=A.dll /r:B.dll -> B
+        ///   3) /r:Foo=A.dll /r:A.dll -> A
+        ///   
+        /// Note that it only makes sense to call this method on a SourceAssemblySymbol since
+        /// alias information is per-compilation.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: It feels strange that the comment mentions SourceAssemblySymbol, the method is on Compilation.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 3]
```diff
@@ -1066,15 +1083,18 @@ internal new NamespaceSymbol GlobalNamespace
...
                 {
                     // Get the root namespace from each module, and merge them all together
-                    HashSet<NamespaceSymbol> allGlobalNamespaces = new HashSet<NamespaceSymbol>();
-                    foreach (ModuleSymbol module in GetAllUnaliasedModules())
-                    {
-                        allGlobalNamespaces.Add(module.GlobalNamespace);
-                    }
 
-                    var result = MergedNamespaceSymbol.Create(new NamespaceExtent(this),
+                    // Get all assemblies in this compilation, including the source assembly and all referenced assemblies.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: We are getting modules, not assemblies.

File: src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs [hunk 4]
```diff
@@ -769,6 +772,7 @@ private NamedTypeSymbol ApplyGenericArguments(NamedTypeSymbol symbol, Type[] typ
                 result = candidate;
             }
 
+            assemblies.Free();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs]: Leaking the builder in the loop above, there are several returns there.

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs [hunk 5]
```diff
@@ -12,6 +12,7 @@
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
+    using Symbols.Retargeting;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: Consider putting usings at the beginning of the file.

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs [hunk 6]
```diff
@@ -227,11 +228,7 @@ public PEAssemblySymbol CreatePEAssemblyForAssemblyMetadata(AssemblyMetadata met
                 // If the compilation has a reference from metadata to source assembly we can't share the referenced PE symbols.
                 Debug.Assert(!HasCircularReference);
 
-                var referencedAssembliesByIdentity = new Dictionary<AssemblyIdentity, AssemblySymbol>();
-                foreach (var symbol in this.ReferencedAssemblies)
-                {
-                    referencedAssembliesByIdentity.Add(symbol.Identity, symbol);
-                }
+                var referencedAssembliesByIdentity = ReferencedAssemblies.ToDictionary(s => s.Identity, s => s);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: I don't see any good reason for changing this code to use delegates.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs (hunk 1 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: It wasn't needed before either, ReferenceAssemblies only contains assemblies, not modules.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs (hunk 2 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: I didn't notice, will fix the comment.

File: src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs (hunk 4 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs]: Fixed.

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs (hunk 5 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: I will.

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs [hunk 7]
```diff
@@ -415,9 +412,12 @@ private bool CreateAndSetSourceAssemblyFullBind(CSharpCompilation compilation)
...
                 // Setup references for the compilation (out parameters)
-                referencedAssembliesMap = new Dictionary<MetadataReference, ReferencedAssembly>(referenceMap.Length);
+                referencedAssembliesMap = new Dictionary<MetadataReference, int>(referenceMap.Length);
                 referencedModulesMap = new Dictionary<MetadataReference, int>(modules.Length);
 
+                var referencedAssemblies = ArrayBuilder<AssemblySymbol>.GetInstance();
+                var aliasesOfReferencedAssemblies = ArrayBuilder<ImmutableArray<string>>.GetInstance();
+
                 var sourceModule = assemblySymbol.SourceModule;
                 var referencedAssemblySymbols = sourceModule.GetReferencedAssemblySymbols();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: It looks like this local is no longer used. Do we really have to rebuild this list in referencedAssemblies builder and allocate new array?

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs (hunk 6 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: It's just shorter.

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs (hunk 7 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: It is used on line 441. I'm not sure if the ordering has to be the same.

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: It is definitely slower. Please revert the change back to the loop.

Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: Sure.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs (hunk 3 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: Comment fixed.

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs [hunk 8]
```diff
@@ -797,24 +804,27 @@ private static void UpdateSymbolCacheNoLock(List<int> newSymbols, ImmutableArray
...
+                protected AssemblyDataForMetadataOrCompilation(
+                    AssemblyIdentity identity,
+                    ImmutableArray<AssemblyIdentity> referencedAssemblies,
+                    bool embedInteropTypes)
                 {
-                    this.EmbedInteropTypes = embedInteropTypes;
+                    Debug.Assert(identity != null);
+
+                    _embedInteropTypes = embedInteropTypes;
+                    _identity = identity;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: _referencedAssemblies are not assigned

Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: Fixed.

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs [hunk 9]
```diff
@@ -1012,128 +978,76 @@ private bool IsMatchingAssembly(PEAssemblySymbol peAssembly)
...
-                    }
-                }
-
-                public override bool DeclaresTheObjectClass
-                {
-                    get
-                    {
-                        return _compilation.DeclaresTheObjectClass;
-                    }
+                    var asm = (candidateAssembly as RetargetingAssemblySymbol)?.UnderlyingAssembly ?? candidateAssembly;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: The logic is not equivalent to the original. Please revert code back to the 'if' statement.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs [hunk 10]
```diff
@@ -98,10 +77,10 @@ public bool DeclarationsAccessibleWithoutAlias()
         private ThreeState _lazyHasCircularReference;
 
         /// <summary>
-        /// A map from a metadata reference to an AssemblySymbol used for it. Do not access
+        /// A map from a metadata reference to an index to <see cref="ReferencedAssemblies"/> array. Do not access
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: Did you mean _lazyReferencedAssemblies array?

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: Please keep comments in terms of fields as they used to be.

File: src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs (hunk 9 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs]: I'll revert. Though it is equivalent since the UnderlyingAssembly can't be null.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs [hunk 11]
```diff
@@ -172,6 +151,11 @@ public bool DeclarationsAccessibleWithoutAlias()
         private ImmutableArray<TAssemblySymbol> _lazyReferencedAssemblies;
 
         /// <summary>
+        /// Assemblies referenced directly by the source module of the compilation.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: How do elements of this array correspond to elements of other arrays.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs (hunk 10 above)
Comment (tmat) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: OK.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs (hunk 11 above)
Comment (tmat) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: Will add a comment.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs [hunk 12]
```diff
@@ -381,45 +380,55 @@ internal bool IsBound
...
+            }
         }
 
-        internal int GetReferencedModuleIndex(MetadataReference reference)
+        public bool DeclarationsAccessibleWithoutAlias(int referencedAssemblyIndex)
         {
-            int index;
-            return ReferencedModuleIndexMap.TryGetValue(reference, out index) ? index : -1;
+            var aliases = AliasesOfReferencedAssemblies[referencedAssemblyIndex];
+            return aliases.Length == 0 || aliases.IndexOf(MetadataReferenceProperties.GlobalAlias) >= 0;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: Should we use explicit comparer with IndexOf

File: src/Compilers/Core/SharedCollections/ArrayBuilder.cs [hunk 13]
```diff
@@ -416,5 +417,41 @@ public void AddMany(T item, int count)
...
+
+            int j = 0;
+            for (int i = 0; i < Count; i++)
+            {
+                if (set.Add(this[i]))
+                {
+                    j++;
+                }
+
+                this[j] = this[i];
```
Comment (AlekseyTs) [inline src/Compilers/Core/SharedCollections/ArrayBuilder.cs]: Where do we truncate the builder? Please add unit-tests for new methods.

File: src/Compilers/Core/SharedCollections/ArrayBuilder.cs [hunk 14]
```diff
@@ -416,5 +417,41 @@ public void AddMany(T item, int count)
...
+            }
+
+            set.Free();
+        }
+
+        public ImmutableArray<S> SelectDistinct<S>(Func<T, S> selector)
+        {
+            var result = ArrayBuilder<S>.GetInstance();
```
Comment (AlekseyTs) [inline src/Compilers/Core/SharedCollections/ArrayBuilder.cs]: Can we use this.Count as a size hint?

File: src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb [hunk 15]
```diff
@@ -187,10 +187,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 ' If the compilation has a reference from metadata to source assembly we can't share the referenced PE symbols.
                 Debug.Assert(Not HasCircularReference)
 
-                Dim referencedAssembliesByIdentity = New Dictionary(Of AssemblyIdentity, AssemblySymbol)()
-                For Each symbol In Me.ReferencedAssemblies
-                    referencedAssembliesByIdentity.Add(symbol.Identity, symbol)
-                Next
+                Dim referencedAssemblySymbols = ReferencedAssemblies
+                Dim referencedAssembliesByIdentity = referencedAssemblySymbols.ToDictionary(Function(s) s.Identity, Function(s) s)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb]: Please revert back to the loop.

File: src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb [hunk 16]
```diff
@@ -361,9 +359,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
                 ' Setup references for the compilation
-                referencedAssembliesMap = New Dictionary(Of MetadataReference, ReferencedAssembly)(referenceMap.Length)
+                referencedAssembliesMap = New Dictionary(Of MetadataReference, Integer)(referenceMap.Length)
                 referencedModulesMap = New Dictionary(Of MetadataReference, Integer)(modules.Length)
 
+                Dim referencedAssemblies = ArrayBuilder(Of AssemblySymbol).GetInstance()
+                Dim aliasesOfReferencedAssemblies = ArrayBuilder(Of ImmutableArray(Of String)).GetInstance()
+
                 Dim sourceModule = assemblySymbol.SourceModule
                 Dim referencedAssemblySymbols = sourceModule.GetReferencedAssemblySymbols()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb]: How is this array different from what we collect in referencedAssemblies. Is it worth allocating another array for what looks like the same data? Perhaps worth a comment then.

File: src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb [hunk 17]
```diff
@@ -421,6 +423,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                     End SyncLock
                 End If
 
+                referencedAssemblies.Free()
+                aliasesOfReferencedAssemblies.Free()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb]: Leaking the builders? There is a return above.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs (hunk 12 above)
Comment (tmat) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs]: Yes! Good catch. I just kept what was there before.

File: src/Compilers/Core/SharedCollections/ArrayBuilder.cs (hunk 13 above)
Comment (tmat) [inline src/Compilers/Core/SharedCollections/ArrayBuilder.cs]: Fixed, added tests.

File: src/Compilers/Core/SharedCollections/ArrayBuilder.cs (hunk 14 above)
Comment (tmat) [inline src/Compilers/Core/SharedCollections/ArrayBuilder.cs]: We could, but in general we don't know how many distinct items are there gonna be.

File: src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb (hunk 15 above)
Comment (tmat) [inline src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb]: Done.

File: src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb (hunk 16 above)
Comment (tmat) [inline src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb]: That's a good question, I'm not sure. Will look into whether we can reuse the existing array.

Comment (tmat) [inline src/Compilers/VisualBasic/Portable/Symbols/ReferenceManager.vb]: Yes, I convinced myself that these are indeed the same arrays and we don't need to rebuild the array. I also run tests with appropriate asserts that verify this and all pass.

**Comment by tmat** [comment]
@VSadov Could you please take a look?

File: src/Compilers/Core/SharedCollections/ArrayBuilder.cs [hunk 18]
```diff
@@ -416,5 +417,42 @@ public void AddMany(T item, int count)
...
+
+        public void RemoveDuplicates()
+        {
+            var set = PooledHashSet<T>.GetInstance();
+
+            int j = 0;
+            for (int i = 0; i < Count; i++)
+            {
+                this[j] = this[i];
+
```
Comment (VSadov) [inline src/Compilers/Core/SharedCollections/ArrayBuilder.cs]: should the assignment be under the if below?

Comment (tmat) [inline src/Compilers/Core/SharedCollections/ArrayBuilder.cs]: Yes, that would be more efficient.

**Comment by VSadov** [comment]
Comment on RemoveDuplicates, otherwise LGTM

**Comment by tmat** [comment]
@VSadov Thanks.
